src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -34,91 +34,164 @@ def self.compile_ast(code, compile_options, boltlib_path)
-
-          # Prior to PE-29443 variables were in a hash. Serialization between ruby/clojure/json did
-          # not preserver hash order necessary for deserialization. The data strucutre is now stored
-          # in a list for moving data and the list is used to construct an ordered ruby hash.
-          plan_variables = if compile_options['variables']['values'].is_a?(Array)
-                        compile_options['variables']['values'].each_with_object({}) do |param_hash, acc|
-                          acc[param_hash.keys.first] = param_hash.values.first
-                        end
-                      else
-                        compile_options['variables']['values']
-                      end
-
-          target_variables = compile_options.dig('target_variables', 'values') || {}
-
-          variables = {
-            variables: plan_variables,
-            target_variables: target_variables,
-          }
-
-          env_conf = {
-            pre_modulepath: boltlib_path,
-            envpath: Puppet[:environmentpath],
-            facts: compile_options['facts']['values'],
+      # Compiling for project plans is the only scenario where we need to understand
+      # if puppetserver is in multithreaded mode (this is because project compiles have
+      # to manage/change puppet's settings to work correctly)
+      def self.compile_for_project_plan(code, compile_options, boltlib_path, multithreaded)kubectl
+        # Modulepath needs to be combined "manually" here because there is no
+        # equivalent to pre_modulepath for in_tmp_environment
+        env_conf = {
+          modulepath: Array(boltlib_path) + Array(compile_options['modulepath']),
+          facts: compile_options['facts']['values'],
+        }
+
+        plan_variables = ordered_plan_vars(compile_options)
+        target_variables = compile_options.dig('target_variables', 'values') || {}
+        variables = {
+          variables: plan_variables,
+          target_variables: target_variables,
+        }
+
+        # TODO: Given we hide this from plan authors this current iteration has only
+        # the "required" data for now. Once we can get https://github.com/puppetlabs/bolt/pull/1770
+        # merged and promoted we can just use an empty hash.
+        fake_config = {
+          'transport' => 'redacted',
+          'transports' => {
+            'redacted' => 'redacted'
           }
+        }
+        bolt_inv = Bolt::ApplyInventory.new(fake_config)
+        bolt_project = Struct.new(:name, :path, :load_as_module?).new(compile_options['project_name'],
+                                                                      compile_options['project_root'],
+                                                                      true)
+        puppet_overrides = {
+          bolt_inventory:  bolt_inv,
+          bolt_project:    bolt_project,
+          node_name_value: compile_options['certname'],
+          hiera_config:    compile_options['hiera_config']
```

Comment:
For some reason it almost looks like `Puppet.override` does not adequately set this. We may just have to do `Puppet.settings[setting]` for node_name_value and hiera_config.